Added support for custom app-boot type in contentFor

dnalagatla opened the pull request:
Fix for #183 

This includes the following:
1. Adds support to use custom `app-boot` of contentFor.
2. Updates that fixes Macro-config error came when building test app with fastboot support. Applied the patch provided by @stefanpenner 
3. Fixed vendor files in Macro-test app.
4. Test that asserts application is booting using custom boot logic added by ember-cli-fastboot addon as part of a separate fastboot app. @stefanpenner.
5. Added test for custom app-boot as part of Macro-test app.
6. Verified by linking to ember app using the ember-cli-fastboot add-on.

cc: @stefanpenner @ef4

stefanpenner commented:
This will depend on us landing: https://github.com/embroider-build/embroider/pull/330 (I'll see if I can wrap that up today)

stefanpenner commented on test-packages/macro-fastboot-tests.test.js:
```diff
@@ -0,0 +1,37 @@
+    const resp = await fastboot.visit('/');
+    expect(resp._body).toContain('<script type="x/boundary" id="fastboot-body-start">');
+    expect(resp._body).toContain('<script type="x/boundary" id="fastboot-body-end">');
+  } catch (e) {
```
This try/catch appears only re-throw something which is already thrown. This leads me to believe it is redundant, can we remove it?

dnalagatla replied:
sure, I will fix that.

dnalagatla replied:
Fixed it.

stefanpenner commented on test-packages/macro-tests/vendor/prepend/four.js:
```diff
@@ -1 +1 @@
-self.ORDER.push('prepend/four');
+self.ORDER && self.ORDER.push('prepend/four');
```
This guard seems to be masking another issue, please ensure the root cause is addressed.

stefanpenner replied:
My existing PR #322 addresses this issue: https://github.com/embroider-build/embroider/pull/322

Specifically, this commit: https://github.com/embroider-build/embroider/pull/322/commits/382d9e5b32748c7a5d6666f1e73596cc8d7158b0

dnalagatla replied:
removed changes and fixed this issue as part o theis PR - https://github.com/embroider-build/embroider/pull/336

stefanpenner commented on test-packages/macro-fastboot-tests.test.js:
```diff
@@ -0,0 +1,37 @@
+    cwd: `${__dirname}/macro-tests`,
+  });
+
+  const buildCompletePromise = new Promise(res => {
```
wouldn't `await execa('ember', ['build'], { ... }` also tell us the build is complete, at which point we know `dirname + '/macro-tests/dists'` is ready to be consumed by fastboot

dnalagatla replied:
I cleaned up the code use await on the build command.

stefanpenner commented on test-packages/macro-fastboot-tests.test.js:
```diff
@@ -0,0 +1,37 @@
+  jest.setTimeout(120000);
+
+  const buildProcess = execa('ember', ['build'], {
+    cwd: `${__dirname}/macro-tests`,
```
This is fine for now, although I know ed would prefer us to create a fresh test-app for fastboot, but if you don't have cycles I can totally help.

dnalagatla replied:
I created a separate fastboot-app for fastboot related testing.

dnalagatla replied:
will be useful for any future changes related to Fastboot add-on

dnalagatla commented:
I just realized, My change for {{content-for app-boot}} didn't go in and the tests were not capturing the correct scenarios.

Also, noticed that changes for supporting app-boot are more involved than I expected, as I cannot update `entryTemplate` in core/app.ts  with an else condition to include custom app-boot content.

https://github.com/embroider-build/embroider/blob/master/packages/core/src/app.ts#L953

My current approach based on my understanding of embroider flow is to:

-  Update entry template to take else condition [here](https://github.com/embroider-build/embroider/blob/master/packages/core/src/app.ts#L953)  

- update the Adapter interface to include `appBoot: string` property. 

- And then update `compat/v1-app.ts` to have a getter for `appBoot` that returns custom  app-boot content when autoRun is false or else return an empty string.

- The `appBoot` getter loops through `this.configReplacePatterns` to get the content for `app-boot` type.

The issue I am seeing is the signature of [compile](https://github.com/embroider-build/embroider/blob/master/packages/core/src/app.ts#L924) doesn't accept custom property and get typescript compilation error. 

Working around ^ compilation issue by making changes to JS code, I can see my approach works. But I am not happy with that.  

 @stefanpenner @ef4 can you recommend a good a way to support `app-boot`.

dnalagatla commented:
@stefanpenner @ef4 I added the changes to support app-boot. As this implementation should be included as part of entryTemplate, I updated the code with following flow:

- Update entry template to take else condition here

- update the Adapter interface to include appBoot: string property.

- And then update compat/v1-app.ts to have a getter for appBoot that returns custom app-boot content when autoRun is false or else return an empty string.

- The appBoot getter logic involves looping through this.configReplacePatterns to get the content for app-boot type.

Please let me know if you have any suggestions.

stefanpenner commented:
@dnalagatla tests appear to be failing, could you take a look?

ef4 commented:
Thanks for updating.

This reimplements config replacement and it doesn't do it the same way the existing config replacement code (in broccoli-config-replace) works. For example, the replacer is not required to be a function, it can also be a string.

Let's continue to use broccoli-config-replace instead of maintaining a separate implementation. There is already a place [here](https://github.com/embroider-build/embroider/blob/117e8451c47257135f17dab975fda1565898118b/packages/compat/src/compat-app.ts#L44) to setup broccoli trees that will be consumed by `CompatAppAdapter`.

Also, this PR still has stray whitespace changes in unrelated files.

dnalagatla commented:
> Thanks for updating.
> 
> This reimplements config replacement and it doesn't do it the same way the existing config replacement code (in broccoli-config-replace) works. For example, the replacer is not required to be a function, it can also be a string.
> 
> Let's continue to use broccoli-config-replace instead of maintaining a separate implementation. There is already a place [here](https://github.com/embroider-build/embroider/blob/117e8451c47257135f17dab975fda1565898118b/packages/compat/src/compat-app.ts#L44) to setup broccoli trees that will be consumed by `CompatAppAdapter`.
> 
> Also, this PR still has stray whitespace changes in unrelated files.

Hi @ef4, thanks for the input. I made the changes based on your suggestion. Thanks to @sparshithNR for helping me with broccoli-tree flow. I have verified my tests are passing.

ef4 commented:
Thank you, nice work.
